src/utils/uintarray-encoding.js

```diff
@@ -4,31 +4,157 @@
+// in this number?" bit. That bit allows us to concatenate multiple encoded numbers
+// without separators, and decoding still knows where to split.
+
+export type UintArrayUrlComponentEncodingType = 'consecutive-12' | 'compact-64';
+
+export function encodeUintSetForUrlComponent(
+  numbers: Set<number>,
+  encodingType: UintArrayUrlComponentEncodingType
+): string {
+  // A set has no order. Convert it to an array and then sort the array,
+  // so that consecutive numbers can be detected by encodeUintArrayConsecutive12.
+  const array = Array.from(numbers);
+  array.sort((a, b) => a - b);
+  return encodeUintArrayForUrlComponent(array, encodingType);
+}
+
+export function encodeUintArrayForUrlComponent(
+  array: number[],
+  encodingType: UintArrayUrlComponentEncodingType
+): string {
+  switch (encodingType) {
+    case 'compact-64':
+      return array.map(uintToStringCompact64).join('');
+    case 'consecutive-12':
+    default:
+      return encodeUintArrayConsecutive12(array);
+  }
+}
+
+export function decodeUintArrayFromUrlComponent(
+  s: string,
+  encodingType: UintArrayUrlComponentEncodingType
+): number[] {
+  switch (encodingType) {
+    case 'compact-64':
+      return decodeUintArrayCompact64(s);
+    case 'consecutive-12':
+    default:
+      return decodeUintArrayConsecutive12(s);
+  }
+}
+
+// Converts [7, 8, 9, 10, 6, 2, 3, 4, 5] to "7_10-6-2_5"
+function encodeUintArrayConsecutive12(numbers: number[]): string {
+  const pieces = [];
+  for (let i = 0; i < numbers.length; i++) {
+    const number = numbers[i];
+    if (i === numbers.length - 1 || numbers[i + 1] !== number + 1) {
+      pieces.push(`${number}`);
+      continue;
+    }
+
+    // The next number is number + 1, so we can create a "range piece".
+    // Check how many consecutive numbers there are.
+    const startNumber = number;
+    let consecutiveCount = 1;
+    while (
+      i + consecutiveCount + 1 < numbers.length &&
+      numbers[i + consecutiveCount + 1] === startNumber + consecutiveCount + 1
+    ) {
+      consecutiveCount++;
+    }
+    const endNumber = startNumber + consecutiveCount;
+    pieces.push(`${startNumber}_${endNumber}`);
+
+    // Skip over the consecutive numbers that we consumed.
+    i += consecutiveCount;
+  }
+  return pieces.join('-');
+}
+
+// Converts "7_10-6-2_5" to [7, 8, 9, 10, 6, 2, 3, 4, 5]
+function decodeUintArrayConsecutive12(s: string): number[] {
+  const pieces = s.split('-');
+  const numbers = [];
+  for (const piece of pieces) {
+    if (piece.includes('_')) {
+      const [fromStr, toInclusiveStr] = piece.split('_');
+      const from = +fromStr;
+      const toInclusive = +toInclusiveStr;
+      if (!isNaN(from) && !isNaN(toInclusive)) {
+        for (let i = from; i <= toInclusive; i++) {
+          numbers.push(i);
+        }
+      } else {
+        console.warn('Unexpected NaN in decodeUintArrayConsecutive12');
+      }
+    } else {
+      numbers.push(+piece);
```

Comment:
I believe this can be NaN as well, so a check + the same warning would be a good idea.